cogs/thread_friends.py

```diff
@@ -0,0 +1,375 @@
+            return set()
+
+        members: set[discord.Member] = {
+            member
+            for member in main_guild.members
+            if not member.bot and guest_role not in member.roles
+        }
+
+        if not ctx.value or ctx.value.startswith("@"):
+            return {
+                discord.OptionChoice(name=f"@{member.name}", value=str(member.id))
+                for member in members
+            }
+
+        return {
+            discord.OptionChoice(name=member.name, value=str(member.id)) for member in members
+        }
+
+    @staticmethod
+    async def autocomplete_get_roles(ctx: "TeXBotAutocompleteContext") -> "AbstractSet[discord.OptionChoice] | AbstractSet[str]":  # noqa: E501
+        """Autocomplete callable that generates the set of available selectable roles."""
+        try:
+            main_guild: discord.Guild = ctx.bot.main_guild
+        except GuildDoesNotExistError:
+            return set()
+
+        if not ctx.value or ctx.value.startswith("@"):
+            return {
+                discord.OptionChoice(name=f"@{role.name}", value=str(role.id))
+                for role in main_guild.roles
+            }
+
+        return {
+            discord.OptionChoice(name=role.name, value=str(role.id))
+            for role in main_guild.roles
+        }
+
+    async def add_user_or_role_silently(self, user_or_role: discord.Member | discord.Role, thread: discord.Thread) -> None:  # noqa: E501
+        """Add a user or role to a thread without pinging them."""
+        message: discord.Message = await thread.send(
+            content=f"Adding {user_or_role!r} to thread...",
+            silent=True,
+        )
+        await message.edit(content=f"{user_or_role.mention}")
+        await message.delete(delay=1)
+
+    async def add_user_or_role_with_ping(self, user_or_role: discord.Member | discord.Role, thread: discord.Thread) -> None:  # noqa: E501
+        """Add a user or role to a thread and ping them."""
+        if isinstance(user_or_role, discord.Member):
+            try:
+                await thread.add_user(user=user_or_role)
+            except discord.NotFound:
+                logger.debug(
+                    "User: %s has blocked the bot and "
+                    "therefore could not be added to thread: %s.",
+                    user_or_role,
+                    thread
+                )
+            return
+
+        if isinstance(user_or_role, discord.Role):
+            member: discord.Member
+            for member in user_or_role.members:
+                try:
+                    await thread.add_user(member)
+                except discord.NotFound:
+                    logger.debug(
+                        "User: %s has blocked the bot and "
+                        "therefore could not be added to thread: %s.",
+                        member,
+                        thread
+                    )
+            return
```

Comment:
* Unnecessary duplication of code
* Use of term "user" not "member"
* Allow for intersecting set of members/roles to be added

```suggestion
    async def add_member_or_role_with_ping(self, *members_and_roles: discord.Member | discord.Role, thread: discord.Thread) -> None:  # noqa: E501
        """Add a user or role to a thread and ping them."""
        members: Iterable[discord.Member] = itertools.chain.from_iterable(
            (member_or_role,)
            if isinstance(member_or_role, discord.Member)
            else member_or_role.members
            for member_or_role in members_and_roles
        )
        
        member: discord.Member
        for member in set(members):
            try:
                await thread.add_user(user=member)
            except discord.NotFound:
                logger.debug(
                    "Member: %s has blocked the bot and "
                    "therefore could not be added to thread: %s",
                    member,
                    thread
                )
```
(Requires import of itertools library)